kong/plugins/rate-limiting/policies/init.lua

```diff
@@ -111,6 +133,84 @@ local function get_redis_connection(conf)
   return red
 end
 
+local function clear_local_counter()
+  table_clear(cur_usage)
+  table_clear(cur_usage_expire_at)
+  table_clear(cur_delta)
+end
+
+local function sync_to_redis(premature, conf)
+  if premature then
+    return
+  end
+
+  local red, err = get_redis_connection(conf)
+  if not red then
+    kong.log.err("[rate-limiting] failed to connect to Redis: ", err)
+    clear_local_counter()
+    return
+  end
+
+  red:init_pipeline()
+
+  for cache_key, delta in pairs(cur_delta) do
+    red:eval([[
+      local key, value, expiration = KEYS[1], tonumber(ARGV[1]), ARGV[2]
+      local exists = redis.call("exists", key)
+      redis.call("incrby", key, value)
+      if not exists or exists == 0 then
+        redis.call("expireat", key, expiration)
+      end
+    ]], 1, cache_key, delta, cur_usage_expire_at[cache_key])
+  end
+
+  local _, err = red:commit_pipeline()
+  if err then
+    kong.log.err("[rate-limiting] failed to commit increment pipeline in Redis: ", err)
+    clear_local_counter()
+    return
+  end
+
+  local ok, err = red:set_keepalive(10000, 100)
+  if not ok then
+    kong.log.err("[rate-limiting] failed to set Redis keepalive: ", err)
+    clear_local_counter()
+    return
+  end
+
+  -- just clear these tables and avoid creating three new tables
+  clear_local_counter()
+end
+
+local function periodical_sync(conf, sync_func)
+  if not auto_sync_timer then
+    local err
+    -- timer may be initialized after the module's loaded so we need to update the reference
+    auto_sync_timer, err = kong.timer:named_every("rate-limiting-auto-sync", conf.sync_rate, sync_func, conf)
+
+    if not auto_sync_timer then
+      kong.log.err("failed to create timer: ", err)
+      return nil, err
+    end
+  end
+
+  return true
+end
+
+local function update_local_counters(conf, periods, limits, identifier, value)
+  for period, period_date in pairs(periods) do
+    if limits[period] then
+      local cache_key = get_local_key(conf, identifier, period, period_date)
+
+      if cur_delta[cache_key] then
+        cur_delta[cache_key] = cur_delta[cache_key] + value
+      else
+        cur_delta[cache_key] = value
+      end
```

Comment:
```suggestion
      if not cur_delta[cache_key] then
         cur_delta[cache_key] = value
      end
      cur_delta[cache_key] = cur_delta[cache_key] + value
```

Reply:
I don't believe the suggested change is equivalent to the original (`cur_delta[cache_key]` would be initialized to `delta * 2` instead of `delta`).  Why the suggestion?